### ISSUE TITLE
Feature Request: Tabulator Custom Editors

### ISSUE THREAD

**Issue by BreadMoirai** [rjaros/kvision#40]
I am requesting support for custom editors in Tabulator as described here.
http://tabulator.info/docs/4.2/edit#edit-custom

**Comment by robert-cronin** [comment]
+1 for this, Tabulator doesn't look like it has a built-in date editor but would be useful for a lot of applications

**Comment by robert-cronin** [comment]
I have tried this with the standard tabulator module and it looks like the modulle will need to be modified a little to allow it. The code I tried looks like the following:
```
val dateEditor = fun(cell: CellComponent, onRendered: dynamic, success: dynamic, cancel: dynamic, editorParams: dynamic): Element {
    //cell - the cell component for the editable cell
    //onRendered - function to call when the editor has been rendered
    //success - function to call to pass the successfuly updated value to Tabulator
    //cancel - function to call to abort the edit and return to a normal cell
    //editorParams - params object passed into the editorParams column definition property

    //create and style editor
    var editor = document.createElement("input")

    editor.setAttribute("type", "date")

    //create and style input
    editor.style.padding = "3px";
    editor.style.width = "100%";
    editor.style.boxSizing = "border-box";

    //Set value of editor to the current value of the cell
    editor.setAttribute("value", Moment(cell.getValue() as String).format("YYYY-MM-DD"))

    //set focus on the select box when the editor is selected (timeout allows for editor to be added to DOM)
    onRendered(fun(){
        editor.focus()
        editor.style.css = "100%"
    })

    //when the value has been set, trigger the cell to update
    fun successFunc(){
        success(Moment(editor.getAttribute("value") ?: "", "YYYY-MM-DD").format("DD/MM/YYYY"))
    }

    editor.addEventListener("change", {successFunc()})
    editor.addEventListener("blur", {successFunc()})

    //return the editor element
    return editor
}
```
Then I call this Anonymous function in the ColumnDefinition:
```
...
ColumnDefinition("DATE", "date", editor = dateEditor, formatter = Formatter.DATETIME),
...
```
But there are a few problems with this approach: 

1. The `Element` class does not have the external `focus()` or `style` modifiers defined so cannot be set here until they are added to KVision `Element`s.
2. The dateEditor above depends on the momentjs library (this is something I've developed in a fork of KVision and will be committed with my next PR).
3. Currently the `editor` parameter does not except dynamic inputs and so cannot accept Anonymous functions.
Hope this helps clarify what needs to be done. If I get time, I might submit a PR to fix these issues (with Robert's blessing of course) since our app would depend on something like this too.

**Comment by rjaros** [comment]
I've already created `formatter` and `formatterFunction` properties in `ColumnDefinition` to allow custom formatters. The same can be done for custom editors - we can just add `editorFunction` next to `editor` property.

**Comment by robert-cronin** [comment]
@BreadMoirai  just for your reference, I'm able to get it working, looks a bit like this:
![image](https://user-images.githubusercontent.com/8540764/57204039-74402480-6fe7-11e9-8574-70456d25a798.png)
Here is the code I used:
```
val dateEditor = fun(cell: CellComponent, onRendered: dynamic, success: dynamic): dynamic {
    //cell - the cell component for the editable cell
    //onRendered - function to call when the editor has been rendered
    //success - function to call to pass the successfuly updated value to Tabulator

    //create and style editor
    val editor = document.createElement("input")

    editor.setAttribute("type", "date")

    //create and style input
    editor.setAttribute("padding", "3px")
    editor.setAttribute("width", "100%")
    editor.setAttribute("boxSizing", "border-box")

    //Set value of editor to the current value of the cell
    editor.setAttribute("value", Moment(cell.getValue() as String).format("YYYY-MM-DD"))

    //when the value has been set, trigger the cell to update
    fun successFunc(){
        success(Moment(editor.getAttribute("value") ?: "", "YYYY-MM-DD").format("DD/MM/YYYY"))
    }

    editor.addEventListener("change", {successFunc()})
    editor.addEventListener("blur", {successFunc()})

    //return the editor element
    return editor
}
.....
 ColumnDefinition("DATE", "date", formatterFunction = dateEditor),
.....
```

**Comment by rjaros** [comment]
I have committed some quite sophisticated code (Kotlin power! :-) that will allow to directly embed most KVision components inside tabulator cells. It works both for formatters and editors, with additional type-safe access to Kotlin data model. There are some glitches, e.g. the `Select` component doesn't work because of click event listeners conflicts, but I've added new `SimpleSelect` component, that should be fine for most use cases. 

I'll put some examples in the showcase project and in the guide after the next realease is published.
